### DISCUSSION TITLE
PGroongaでのNormalizer指定をいかにPGroonga自身で調べるか？

### DISCUSSION THREAD

**Discussion by askdkc** [groonga/groonga#1586] [グルカイ！] [answered]
## PGroongaから検索対象のカラムのNormalizerをチェックする方法が気になります

こちらのIssueに関連して

https://github.com/pgroonga/pgroonga/issues/278

「あれ？これ検索実行時にPGroongaがNormalizerを検知する方法ってあるの？」と疑問を持ったので質問します

## 自分の頭の中の想像（C言語の理解が甘いのでかなり適当）

PGroongaではインデックス作成時にNormalizerが指定されていればそれを使い、特に指定がなければ`NormalizerAuto`が使われる
↓
恐らく指定されたNormalizer情報でgrn_objとかにノーマライズされたデータを放り込む（そのためGroongaでは既にノーマライズされた情報のみが入っているから、grn_objからは使用しているNormalizer情報が取れない：という勝手な想像）
↓
このIssueを上手く解決するには`LIKE`演算子で検索時に該当のカラムが`NormalizerAuto`なのか、それ以外かをチェックして、デフォルトの`NormalizerAuto`の場合はrecheckをfalseにすれば良い
↓
でも、PGroongaから該当のカラム検索時には既に何もかもノーマライズ済みで、どこにもNormalizerの情報は入って来ないから、そんな情報は取れない
↓
この場合、PGroongaからPostgreSQLに機能を使って該当カラムのインデックスを調べて、そこから適用されてるNormalizerを調べた方が良い？そしてrecheckをfalseにできる？

**Comment by askdkc** [comment]
## 今回の動画

https://www.youtube.com/live/VmdEuO60jpo

### 回答
grn_objにNormalizer情報は有り(動画で取得方法については説明されてなかったので、具体的な方法は不明。でもオブジェクト内にはあるよ〜と言うことで)

Issueの対応については難しく考えずに、ユーザにインデックス作成時に指定させて、それを見てrecheckをfalseにするだけで良し👍

**Comment by kou** [comment, marked as answer]
https://github.com/pgroonga/pgroonga/blob/f24bb81e02a7fbfecbb5f85d8e04172152c69be1/src/pgroonga.c#L7329C3-L7330 の`GRN_INFO_DEFAULT_TOKENIZER`を`GRN_INFO_NORMALIZER`にすれば取得できます！
